scte35/state.go

```diff
@@ -0,0 +1,115 @@
+/*
+MIT License
+
+Copyright 2016 Comcast Cable Communications Management, LLC
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package scte35
+
+import "github.com/comcast/gots"
+
+type state struct {
+	open        []SegmentationDescriptor
+	blackoutIdx int
+	inBlackout  bool
+}
+
+func NewState() State {
+	return &state{}
+}
+
+func (s *state) Open() []SegmentationDescriptor {
+	open := make([]SegmentationDescriptor, len(s.open))
+	copy(open, s.open)
+	if s.inBlackout {
+		return append(open[0:s.blackoutIdx], open[s.blackoutIdx+1:]...)
+	} else {
+		return open
+	}
+}
+
+func (s *state) ProcessDescriptor(desc SegmentationDescriptor) ([]SegmentationDescriptor, error) {
+	var err error
+	var closed []SegmentationDescriptor
+	// check if this is a duplicate
+	for _, d := range s.open {
+		if desc.Equal(d) {
+			return nil, mpegts.ErrSCTE35DuplicateDescriptor
+		}
+	}
+	// first close signals until one returns false, then handle the breakaway
+	for i := len(s.open) - 1; i >= 0; i-- {
+		d := s.open[i]
+		if desc.CanClose(d) {
+			closed = append(closed, d)
+		} else {
+			break
+		}
+	}
+	// remove all closed descriptors
+	s.open = s.open[0 : len(s.open)-len(closed)]
+	if desc.IsOut() {
+		s.open = append(s.open, desc)
+	} else {
+		// check for validity
+		switch desc.TypeID() {
+		case SegDescProgramEnd,
```

Comment:
Is this right?  Can we be certain that this will be true?  With the case of out-in-in, the first in does not cause the original out to be closed.